### ISSUE TITLE
Create tutorial for extension writers

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor.org#347]
Create a tutorial intended for extension writers that steps through the process of creating an extension. After a general introduction to the extension mechanism, the tutorial should cover on each type of extension. The tutorial should also explain how extensions are loaded and, ideally, how to package extensions so that they can be shared.

**Comment by graphitefriction** [comment]
Let's see if we can outline this at Devoxx FR or Mix IT.

**Comment by graphitefriction** [comment]
Draft outline for Extension Tutorial per conversation with @mojavelinux 

Ruby Extension Tutorial

.Overview of "The Extension"
. What is an extension?
. What are the powers of extensions?
.. Tell them in plain english what they can do (may also be a diagram)
.. Explain that Asciidoctor is extensible
.. Various ways to modify it
... Tap into AsciiDoc
... Effect on the parsing (high level explanation)
... introduce new syntax
... handle include directives
... contribute new docinfo hooks
... preprocess text
... modify parsed tree
... have custom converter
... modify includes

.Step through an existing extension
. Use a basic extension, such as a block or block macro extension, as the example.
. Step through an existing extension, explaining parts and behavior

.Your extension
. Create extension (Focus on creating just one type of extension, such as a block/block macro/gist)
. Register extension
. Load extension
. Possibly package extension (or link out to another tutorial explaining packaging)

.Extension types
. List and define other types of extensions
.Repeat creation, register, and loading of extension for each type

.Nodes and the AST
. Creating nodes in the AST tree
.. A lot of extensions return a node

Also need a parallel guide for writing extensions in Java and JavaScript <- but not very defined yet/not ready for general availability

**Comment by mojavelinux** [comment]
The general idea, as outlined above, is to not slam the reader up front with all the different examples of extensions, but rather give them a general understanding of how AsciiDoc can be extended using the Asciidoctor extension mechanism (awareness), given the a simple example to start with from end-to-end (quickstart), then go through each of the extensions in turn (depth).

We can have a parallel tutorial for writing extension in Java and Groovy, but let's start with Ruby and then we can repurpose the content as needed.
